File: Java/001. Two Sum/twoSum.java
'''
Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

Example:

Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
'''


class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer,Integer> mp = new HashMap<>();
        for(int i=0;i<nums.length;i++){
            int part = target-nums[i];
            if(mp.containsKey(part)){
                return new int[] {mp.get(part), i};
            }
            else{
                mp.put(nums[i],i);
            }
        }
        return null;           
    }
}

File: python/TwoSum.py
'''
Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
'''

#element in set,dict 通过hash查值,速度快
#i, element in enumerate(object) 枚举，获取角标和元素，角标默认从0开始，可指定
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        d = {}
        for i, element in enumerate(nums):
            if target-element in d:
                return [d[target-element], i]
            d[element] = i
